### PR TITLE
Naive text diff Checker utility for Scalasigs (fix #463)

### LINKED ISSUE #463: Text-based Checkscalasig
Builds to #243. Just need some naive-ish working text-diff implementation for now.

### PR DESCRIPTION
**Pull request by wiwa**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -61,12 +61,12 @@ lazy val check = project
     moduleName := "rsc-check",
     scalacOptions += "-Xexperimental",
     libraryDependencies += "ch.qos.logback" % "logback-classic" % "1.2.3",
-    libraryDependencies += "io.github.java-diff-utils" % "java-diff-utils" % "4.0",
     libraryDependencies += "commons-cli" % "commons-cli" % "1.4",
     libraryDependencies += "org.ow2.asm" % "asm" % V.asm,
     libraryDependencies += "org.ow2.asm" % "asm-util" % V.asm,
     libraryDependencies += "org.scala-lang" % "scalap" % V.scala,
     libraryDependencies += "org.eclipse.jgit" % "org.eclipse.jgit" % "4.6.0.201612231935-r",
+    libraryDependencies += "io.github.java-diff-utils" % "java-diff-utils" % "4.0",
```
Comment (ShaneDelmore) [inline build.sbt]: I don't mind the re-order here, but what is the reason behind it?  Looks like it's moved down to be between org.scala-lang.scalap and org.scala-lang.scala-compiler.  Did you mean to sort them all?

File: check/src/main/scala/rsc/checkscalasig/Settings.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
+package rsc.checkscalasig
+
+import java.io.File.pathSeparator
+import java.nio.file.{Path, Paths}
+import rsc.checkbase.SettingsBase
+
+final case class Settings(
+    cp: List[Path] = Nil,
+    ins: List[Path] = Nil,
```
Comment (ShaneDelmore) [inline check/src/main/scala/rsc/checkscalasig/Settings.scala]: What is ins?

Comment (ShaneDelmore) [reply check/src/main/scala/rsc/checkscalasig/Settings.scala]: Looking at the code it appears to be a list of input paths.

File: check/src/main/scala/rsc/checkscalasig/Settings.scala [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+) extends SettingsBase
+
+object Settings {
+  def parse(args: List[String]): Either[List[String], Settings] = {
+    def loop(
+        settings: Settings,
+        allowOptions: Boolean,
+        args: List[String]): Either[List[String], Settings] = {
+      args match {
+        case "--" +: rest =>
```
Comment (ShaneDelmore) [inline check/src/main/scala/rsc/checkscalasig/Settings.scala]: I would prefer using an arg parsing library here both to show usage and for readability.  I can't easily tell what the use case for passing -- as an arg is but it appears to short circuit later options.

File: check/src/main/scala/rsc/checkscalasig/Settings.scala [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+          loop(settings, false, rest)
+        case "--classpath" +: s_cp +: rest if allowOptions =>
+          val cp = s_cp.split(pathSeparator).map(s => Paths.get(s)).toList
+          loop(settings.copy(cp = settings.cp ++ cp), true, rest)
+        case "--quiet" +: rest if allowOptions =>
+          loop(settings.copy(quiet = true), true, rest)
+        case flag +: rest if allowOptions && flag.startsWith("-") =>
+          Left(List(s"unknown flag $flag"))
```
Comment (ShaneDelmore) [inline check/src/main/scala/rsc/checkscalasig/Settings.scala]: An example invocation would be helpful, especially if we hit this use case which could be a typo, or could mean the user doesn't understand how to invoke the tool. Maybe something like: 
```
Left(List(
  s"unknown flag $flag",
  "example invocation: foo --bar -baz"
))
```

File: scalasig/scalasig/src/main/scala/scala/meta/scalasig/Binaries.scala [hunk 5]
```diff
@@ -59,4 +59,13 @@ object Binaries {
       }
     }
   }
+
+  def list(paths: Path*): List[Binary] = {
+    var ress = List.empty[Binary]
```
Comment (ShaneDelmore) [inline scalasig/scalasig/src/main/scala/scala/meta/scalasig/Binaries.scala]: If you expect this List to be large enough for performance to be a concern a ListBuffer would be the higher performance solution.
```
    val ress = ListBuffer.empty[Binary]

     apply(paths.toList) { res =>
      ress += res
    }
    ress.result
```

File: check/src/main/scala/rsc/checkscalasig/Settings.scala (hunk 3 above)
Comment (wiwa) [reply check/src/main/scala/rsc/checkscalasig/Settings.scala]: That will be part of a larger migration to a real parsing library.

**Pull request merged**